candidate_base.go

```diff
@@ -206,57 +205,53 @@ func (c *candidateBase) start(a *Agent, conn net.PacketConn, initializedCh <-cha
 }
 
 func (c *candidateBase) recvLoop(initializedCh <-chan struct{}) {
-	defer func() {
-		close(c.closedCh)
-	}()
+	defer close(c.closedCh)
 
 	select {
 	case <-initializedCh:
 	case <-c.closeCh:
 		return
 	}
 
-	log := c.agent().log
-	buffer := make([]byte, receiveMTU)
+	buf := make([]byte, receiveMTU)
 	for {
-		n, srcAddr, err := c.conn.ReadFrom(buffer)
+		n, sAddr, err := c.conn.ReadFrom(buf)
 		if err != nil {
 			return
 		}
 
-		handleInboundCandidateMsg(c, c, buffer[:n], srcAddr, log)
+		c.handleInboundPacket(buf[:n], sAddr)
 	}
 }
 
-func handleInboundCandidateMsg(ctx context.Context, c Candidate, buffer []byte, srcAddr net.Addr, log logging.LeveledLogger) {
-	if stun.IsMessage(buffer) {
+func (c *candidateBase) handleInboundPacket(buf []byte, sAddr net.Addr) {
+	a := c.agent()
+
+	if stun.IsMessage(buf) {
 		m := &stun.Message{
-			Raw: make([]byte, len(buffer)),
+			Raw: make([]byte, len(buf)),
 		}
 		// Explicitly copy raw buffer so Message can own the memory.
-		copy(m.Raw, buffer)
+		copy(m.Raw, buf)
 		if err := m.Decode(); err != nil {
-			log.Warnf("Failed to handle decode ICE from %s to %s: %v", c.addr(), srcAddr, err)
+			a.log.Warnf("Failed to handle decode ICE from %s to %s: %v", c.addr(), sAddr, err)
 			return
 		}
-		err := c.agent().run(ctx, func(ctx context.Context, agent *Agent) {
-			agent.handleInbound(m, c, srcAddr)
+		err := a.run(c, func(ctx context.Context, a *Agent) {
+			a.handleInbound(m, c, sAddr)
 		})
 		if err != nil {
-			log.Warnf("Failed to handle message: %v", err)
+			a.log.Warnf("Failed to handle message: %v", err)
+		}
+	} else {
+		if a.validateNonSTUNTraffic(c, sAddr) { //nolint:contextcheck
+			// Note: This will return packetio.ErrFull if the buffer ever manages to fill up.
+			if _, err := a.buffer.Write(buf); err != nil {
+				a.log.Warnf("Failed to write packet: %s", err)
+			}
+		} else {
+			a.log.Warnf("Discarded message from %s, not a valid remote candidate", c.addr())
 		}
-
```

Comment:
[Effective Go](https://go.dev/doc/effective_go#if) recommends early return over nesting.

If you feel strongly happy to leave! Personally I think it reduces the complexity of reading

Reply:
Agreed. I reverted the nested flow.